internal/beater/otlp/http.go

```diff
@@ -44,28 +50,121 @@ func init() {
+func NewHTTPHandlers(processor model.BatchProcessor) HTTPHandlers {
 	// TODO(axw) stop assuming we have only one OTLP HTTP consumer running
 	// at any time, and instead aggregate metrics from consumers that are
 	// dynamically registered and unregistered.
 	consumer := &otel.Consumer{Processor: processor}
 	httpMonitoredConsumer.set(consumer)
+	return HTTPHandlers{consumer: consumer}
+}
+
+// HTTPHandlers encapsulates http.HandlerFuncs for handling traces, metrics, and logs.
+type HTTPHandlers struct {
+	consumer *otel.Consumer
+}
+
+// HandleTraces is an http.HandlerFunc that receives a protobuf-encoded traces export
+// request, and processes it with the handler's OTLP consumer.
+func (h HTTPHandlers) HandleTraces(w http.ResponseWriter, r *http.Request) {
+	req := ptraceotlp.NewExportRequest()
+	if err := h.readRequest(r, req); err != nil {
+		h.writeError(w, err, http.StatusBadRequest)
+		return
+	}
+	if err := h.consumer.ConsumeTraces(r.Context(), req.Traces()); err != nil {
+		h.writeError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if err := h.writeResponse(w, ptraceotlp.NewExportResponse()); err != nil {
+		h.writeError(w, err, http.StatusInternalServerError)
+		return
+	}
+}
+
+// HandleMetrics is an http.HandlerFunc that receives a protobuf-encoded metrics export
+// request, and processes it with the handler's OTLP consumer.
+func (h HTTPHandlers) HandleMetrics(w http.ResponseWriter, r *http.Request) {
+	req := pmetricotlp.NewExportRequest()
+	if err := h.readRequest(r, req); err != nil {
+		h.writeError(w, err, http.StatusBadRequest)
+		return
+	}
+	if err := h.consumer.ConsumeMetrics(r.Context(), req.Metrics()); err != nil {
+		h.writeError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if err := h.writeResponse(w, pmetricotlp.NewExportResponse()); err != nil {
+		h.writeError(w, err, http.StatusInternalServerError)
+		return
+	}
+}
+
+// HandleLogs is an http.HandlerFunc that receives a protobuf-encoded logs export
+// request, and processes it with the handler's OTLP consumer.
+func (h HTTPHandlers) HandleLogs(w http.ResponseWriter, r *http.Request) {
+	req := plogotlp.NewExportRequest()
+	if err := h.readRequest(r, req); err != nil {
+		h.writeError(w, err, http.StatusBadRequest)
+		return
+	}
+	if err := h.consumer.ConsumeLogs(r.Context(), req.Logs()); err != nil {
+		h.writeError(w, err, http.StatusInternalServerError)
+		return
+	}
+	if err := h.writeResponse(w, plogotlp.NewExportResponse()); err != nil {
+		h.writeError(w, err, http.StatusInternalServerError)
+		return
+	}
+}
+
+type protoUnmarshaler interface {
+	UnmarshalProto([]byte) error
+}
+
+type protoMarshaler interface {
+	MarshalProto() ([]byte, error)
+}
 
-	tracesHandler, err := otlpreceiver.TracesHTTPHandler(context.Background(), consumer)
+func (h HTTPHandlers) readRequest(req *http.Request, out protoUnmarshaler) error {
+	body, err := io.ReadAll(req.Body)
```

Comment:
I think we should close the `req.Body` here.

Reply:
Why? The net/http server will close the request body once the handler returns.

Reply:
Ah yes, you are right.